File: Examples/iOS/TouchRegions/TouchRegions/FMOscillator.m
//
//  FMOscillator.m
//  TouchRegions
//
//  Created by Aurelius Prochazka on 8/7/14.
//  Copyright (c) 2014 Aurelius Prochazka. All rights reserved.
//

#import "FMOscillator.h"

@implementation FMOscillator

- (instancetype)init {
    self = [super init];
    if (self) {
        
        // INSTRUMENT CONTROL ==================================================
        _frequency            = [[AKInstrumentProperty alloc] initWithValue:440 minimum:1.0 maximum:880];
        _carrierMultiplier    = [[AKInstrumentProperty alloc] initWithValue:1.0 minimum:0.0 maximum:2.0];
        _modulatingMultiplier = [[AKInstrumentProperty alloc] initWithValue:1.0 minimum:0.0 maximum:2.0];
        _modulationIndex      = [[AKInstrumentProperty alloc] initWithValue:15  minimum:0   maximum:30];
        _amplitude            = [[AKInstrumentProperty alloc] initWithValue:0.0 minimum:0   maximum:0.2];
        
        [self addProperty:_frequency];
        [self addProperty:_carrierMultiplier];
        [self addProperty:_modulatingMultiplier];
        [self addProperty:_modulationIndex];
        [self addProperty:_amplitude];
        
        // INSTRUMENT DEFINITION ===============================================
        
        AKFMOscillator *fmOscillator;
        fmOscillator = [[AKFMOscillator alloc] initWithFunctionTable:[AKManager standardSineWave]
                                                       baseFrequency:_frequency
                                                   carrierMultiplier:_carrierMultiplier
                                                modulatingMultiplier:_modulatingMultiplier
                                                     modulationIndex:_modulationIndex
                                                           amplitude:_amplitude];
        [self connect:fmOscillator];
        
        // AUDIO OUTPUT ========================================================
        
        AKAudioOutput *audio = [[AKAudioOutput alloc] initWithAudioSource:fmOscillator];
        [self connect:audio];
    }
    return self;
}

@end

File: AudioKit/Common/Operations/Generators/Oscillators/fmOscillator.swift
//
//  fmOscillator.swift
//  AudioKit
//
//  Created by Aurelius Prochazka, revision history on Github.
//  Copyright © 2016 AudioKit. All rights reserved.
//

import Foundation

extension AKOperation {

    /// Classic FM Synthesis audio generation.
    ///
    /// - Parameters:
    ///   - baseFrequency: In cycles per second, or Hz, this is the common denominator for the carrier and modulating frequencies. (Default: 440, Minimum: 0.0, Maximum: 20000.0)
    ///   - carrierMultiplier: This multiplied by the baseFrequency gives the carrier frequency. (Default: 1.0, Minimum: 0.0, Maximum: 1000.0)
    ///   - modulatingMultiplier: This multiplied by the baseFrequency gives the modulating frequency. (Default: 1.0, Minimum: 0.0, Maximum: 1000.0)
    ///   - modulationIndex: This multiplied by the modulating frequency gives the modulation amplitude. (Default: 1.0, Minimum: 0.0, Maximum: 1000.0)
    ///   - amplitude: Output Amplitude. (Default: 0.5, Minimum: 0.0, Maximum: 10.0)
    ///
    public static func fmOscillator(
        baseFrequency baseFrequency: AKParameter = 440,
        carrierMultiplier: AKParameter = 1.0,
        modulatingMultiplier: AKParameter = 1.0,
        modulationIndex: AKParameter = 1.0,
        amplitude: AKParameter = 0.5
        ) -> AKOperation {
            return AKOperation(module: "fm",
                               inputs: baseFrequency, amplitude,
                                       carrierMultiplier, modulatingMultiplier, modulationIndex)
    }
}
